packages/core/src/api/nodeConversions/nodeToBlock.ts

```diff
@@ -488,3 +492,353 @@ export function nodeToBlock<
+  to: number,
+  schema: Schema,
+  blockSchema: BSchema = getBlockSchema(schema) as BSchema,
+  inlineContentSchema: I = getInlineContentSchema(schema) as I,
+  styleSchema: S = getStyleSchema(schema) as S
+) {
+  const { tr } = addSelectionMarkersTr(state, from, to);
+  return docToBlocks(
+    tr.doc,
+    schema,
+    blockSchema,
+    inlineContentSchema,
+    styleSchema
+  );
+}
+
+/**
+ * Add selection markers to the document at the given positions and return the blocks that span the selection.
+ */
+export function getSelectedBlocksWithSelectionMarkers<
+  BSchema extends BlockSchema,
+  I extends InlineContentSchema,
+  S extends StyleSchema
+>(
+  state: EditorState,
+  from: number,
+  to: number,
+  schema: Schema,
+  blockSchema: BSchema = getBlockSchema(schema) as BSchema,
+  inlineContentSchema: I = getInlineContentSchema(schema) as I,
+  styleSchema: S = getStyleSchema(schema) as S
+) {
+  const { tr, newEnd } = addSelectionMarkersTr(state, from, to);
+
+  return getBlocksBetween(
+    from,
+    newEnd,
+    tr.doc,
+    schema,
+    blockSchema,
+    inlineContentSchema,
+    styleSchema
+  );
+}
+
+/**
+ * Returns all blocks between two positions in a document, but without automatically including parent blocks
+ */
+export function getBlocksBetween<
+  BSchema extends BlockSchema,
+  I extends InlineContentSchema,
+  S extends StyleSchema
+>(
+  start: number,
+  end: number,
+  doc: Node,
+  schema: Schema,
+  blockSchema: BSchema = getBlockSchema(schema) as BSchema,
+  inlineContentSchema: I = getInlineContentSchema(schema) as I,
+  styleSchema: S = getStyleSchema(schema) as S,
+  blockCache = getBlockCache(schema)
+) {
+  const startPosInfo = getNearestBlockPos(doc, start);
+  const endPosInfo = getNearestBlockPos(doc, end);
+  const startNode = getBlockInfo(startPosInfo);
+  const endNode = getBlockInfo(endPosInfo);
+
+  const slice = doc.slice(
+    startNode.bnBlock.beforePos,
+    endNode.bnBlock.afterPos,
+    true
+  );
+
+  const bnSelection = prosemirrorSliceToSlicedBlocks(
+    slice,
+    schema,
+    blockSchema,
+    inlineContentSchema,
+    styleSchema,
+    blockCache
+  );
+
+  if (bnSelection.blockCutAtEnd || bnSelection.blockCutAtStart) {
+    throw new Error("unexpected content cut in getBlocksBetween");
+  }
+
+  // we don't care about the slice metadata, because our slice is based on complete blocks, the
+  return bnSelection.blocks;
+}
+
+/**
+ *
+ * Parse a Prosemirror Slice into a BlockNote selection. The prosemirror schema looks like this:
+ *
+ * <blockGroup>
+ *   <blockContainer> (main content of block)
+ *       <p, heading, etc.>
+ *   <blockGroup> (only if blocks has children)
+ *     <blockContainer> (child block)
+ *       <p, heading, etc.>
+ *     </blockContainer>
+ *    <blockContainer> (child block 2)
+ *       <p, heading, etc.>
+ *     </blockContainer>
+ *   </blockContainer>
+ *  </blockGroup>
+ * </blockGroup>
+ *
+ */
+export function prosemirrorSliceToSlicedBlocks<
```

Comment:
@matthewlipski I think this code is in part similar to some of the logic you wrote in getSelection

Can you take an extra look at this and see if it makes sense - or even whether we should combine the two somehow? (don't spend > 1 hour investigating)